core/pta/attestation.c

```diff
@@ -0,0 +1,589 @@
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[2] << 8) | att_data.pub[3];
+	if ((size + 4) != att_data.pub_size)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding first integer aka modulus
+	if (att_data.pub[4] != 0x02 || att_data.pub[5] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[6] << 8) | att_data.pub[7];
+	*n = &att_data.pub[8];
+	*n_size = size;
+
+	// Decoding second integer aka exponent
+	if (att_data.pub[size + 8] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	*e_size = att_data.pub[size + 9];
+	*e = &att_data.pub[size + 10];
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak get_sys_measurement(uint8_t *ptr)
+{
+	// TODO: Dummy function
+	// get_sys_measurement -> do something hardware specific
+	uint8_t sys[] = {0x00, 0x01, 0x02, 0x03,
+			0x04, 0x05, 0x06, 0x07,
+			0x08, 0x09, 0x0a, 0x0b,
+			0x0c, 0x0d, 0x0e, 0x0f,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00 };
+	memcpy(ptr, sys, 32);
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak get_endorsement_key(uint8_t *key)
+{
+	// TODO: Dummy function get_endorsement_key
+	// do something hardware specific
+	uint8_t end_key[] = {0x9a, 0x04, 0xaa, 0x18,
+			0x2d, 0x03, 0x96, 0x74,
+			0x70, 0x8c, 0xe8, 0x07,
+			0xed, 0x91, 0x4c, 0xd1,
+			0x53, 0xcd, 0x9d, 0xf7,
+			0x80, 0x5e, 0x61, 0x74,
+			0x2f, 0x0a, 0xe4, 0x12,
+			0x94, 0x75, 0x8d, 0xd3 };
+	memcpy(key, end_key, 32);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result decrypt_priv_key(uint8_t *plain, size_t *plain_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *ctx = NULL;
+	size_t block_size = 0;
+	uint32_t algo = TEE_ALG_AES_CBC_NOPAD;
+	uint32_t mode = TEE_MODE_DECRYPT;
+	size_t aes_key_len = TEE_AES_MAX_KEY_SIZE;
+	uint8_t endorsement_key[aes_key_len];
+	uint8_t *dest_ptr = plain;
+	uint8_t *src_ptr = att_data.enc_priv;
+	uint32_t enc_size = att_data.enc_priv_size;
+
+	res = crypto_cipher_get_block_size(algo, &block_size);
+	if (res)
+		return res;
+
+	if (enc_size % block_size != 0)
+		return TEE_ERROR_GENERIC;
+
+	res = get_endorsement_key(endorsement_key);
+	if (res)
+		return res;
+
+	res = crypto_cipher_alloc_ctx(&ctx, algo);
+	if (res)
+		return res;
+
+	res = crypto_cipher_init(ctx, mode, endorsement_key,
+				 aes_key_len, NULL, 0,
+				 att_data.iv, att_data.iv_size);
+	if (res)
+		return res;
+
+	for (size_t i = 0; i < enc_size / block_size; i++) {
+		res = crypto_cipher_update(ctx, mode, false,
+					   src_ptr, block_size, dest_ptr);
+		if (res)
+			return res;
+		dest_ptr += block_size;
+		src_ptr += block_size;
+	}
+
+	crypto_cipher_final(ctx);
+	*plain_size = enc_size - plain[enc_size - 1];
+
+	return res;
+}
+
+static TEE_Result get_sha256(uint8_t *dst, uint8_t *src, size_t src_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *ctx = NULL;
+	size_t digest_len = TEE_SHA256_HASH_SIZE;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		return res;
```

Comment:
Memory leakage, `ctx` isn't freed.

Reply:
I do think, I free it with `crypto_hash_free_ctx(ctx);` on line 239.

Reply:
Not when returning here.

Reply:
Oh, I got it. That is the reason for this `goto err:` structure you are using in the core.

Reply:
Fix [here](https://github.com/OP-TEE/optee_os/pull/4011/commits/c3e8a401eb4aad9a07affde8d829d63979b9f53a)